### PR TITLE
Split logger for charms

### PR DESCRIPTION
**Pull request by SimonRichardson**
The logger in the charmhub and charms facade packages require a
cyclic dependency on the interface. This causes lots of problems
when implementing the logger. To make matters worse, the interface
also had a loggo.Logger attached to some loggers, which makes it
impossible to satisfy that interface if you've got a sub-type of that
interface.

The solution is to make two interfaces a logger factory and a logger.
This removes the cyclic nature of the interface and cleans up what
the actual dependency is.

This is a follow on from PRs: 
 - https://github.com/juju/juju/pull/16672
 - https://github.com/juju/juju/pull/16681
 - https://github.com/juju/juju/pull/16683

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test --build-agent
```

In a different terminal, run `juju debug-log -m controller`

Then run the following to see it all still works.

```sh
$ juju model-config -m controller logging-config="<root>=INFO;#charmhub=TRACE;#http=TRACE"
$ juju add-model default
$ juju deploy ubuntu
```

## Links

**Jira card:** JUJU-5137

### CODE REVIEW

File: worker/bootstrap/worker_test.go [hunk 1]
```diff
@@ -129,5 +130,5 @@ func (s *workerSuite) expectObjectStoreGetter() {
 }
 
 func (s *workerSuite) expectBootstrapFlagSet() {
-	s.flagService.EXPECT().SetFlag(gomock.Any(), BootstrapFlag, true).Return(nil)
+	s.flagService.EXPECT().SetFlag(gomock.Any(), flags.BootstrapFlag, true).Return(nil)
 }
```
Comment (SimonRichardson) [inline worker/bootstrap/worker_test.go:134]: Driveby: this was a mistake to use the constant in a worker.

**Comment by hpidcock** [comment]
Starting to feel like we should adopt slog.

File: internal/charmhub/client.go [hunk 2]
```diff
@@ -103,11 +92,11 @@ type Client struct {
 
 // NewClient creates a new Charmhub client from the supplied configuration.
 func NewClient(config Config) (*Client, error) {
-	logger := config.Logger
-	if logger == nil {
+	loggerFactory := config.LoggerFactory
+	if loggerFactory == nil {
 		return nil, errors.NotValidf("nil logger")
```
Comment (manadart) [inline internal/charmhub/client.go:97]: "nil logger factory"

File: core/base/base.go [hunk 3]
```diff
@@ -21,6 +21,9 @@ type Base struct {
 	Channel Channel
 }
 
+// Empty is an empty base.
+var Empty = Base{}
```
Comment (manadart) [inline core/base/base.go:25]: This doesn't appear to be used.

File: apiserver/facades/client/charms/services/downloader.go [hunk 4]
```diff
@@ -68,3 +68,25 @@ func (s repoFactoryShim) GetCharmRepository(ctx context.Context, src corecharm.S
 
 	return repo, err
 }
+
+type loggoLoggerFactory struct {
+	Logger loggo.Logger
```
Comment (manadart) [inline apiserver/facades/client/charms/services/downloader.go:73]: I would embed `charmhub.LoggerFactory` here instead of duplicating `Child` and `ChildWithLabels`.

**Review by manadart — COMMENTED** [review]
Some suggestions.

File: apiserver/facades/client/charms/services/downloader.go (hunk 4 above)
Comment (SimonRichardson) [reply apiserver/facades/client/charms/services/downloader.go:73]: I'll see what I can do with this one, it was painful because of invariance the last time I tried.

Comment (SimonRichardson) [reply apiserver/facades/client/charms/services/downloader.go:73]: Bitten by invariance.

**Review by manadart — APPROVED** [review]
Thanks for the changes. All working well.
